packages/core/src/types.ts

```diff
@@ -0,0 +1,371 @@
+/**
+ * @fileoverview Shared types for ESLint Core.
+ */
+
+//------------------------------------------------------------------------------
+// Helper Types
+//------------------------------------------------------------------------------
+
+/**
+ * Represents an error inside of a file.
+ */
+export interface FileError {
+	message: string;
+	line: number;
+	column: number;
+	endLine?: number;
+	endColumn?: number;
+}
+
+/**
+ * Represents a problem found in a file.
+ */
+export interface FileProblem {
+	ruleId: string | null;
+	message: string;
+	loc: SourceLocation;
+}
+
+//------------------------------------------------------------------------------
+// ASTs
+//------------------------------------------------------------------------------
+
+/**
+ * Represents an AST node or token with location information.
+ */
+export interface SyntaxElement {
+	loc: SourceLocation;
+	range?: [number, number];
+	[key: string]: any;
+}
+
+/**
+ * Represents the start and end coordinates of a node inside the source.
+ */
+export interface SourceLocation {
+	start: Position;
+	end: Position;
+}
+
+/**
+ * Represents a location coordinate inside the source.
+ */
+export interface Position {
+	line: number;
+	column: number;
+	offset: number;
+}
+
+//------------------------------------------------------------------------------
+// Config
+//------------------------------------------------------------------------------
+
+export type RuleSeverity = 0 | 1 | 2 | "off" | "warn" | "error";
+
+export type RuleConfig = RuleSeverity | [RuleSeverity, ...any[]];
+
+export interface RulesConfig {
+	[key: string]: RuleConfig;
+}
+
+//------------------------------------------------------------------------------
+// Languages
+//------------------------------------------------------------------------------
+
+/**
+ * Represents a plugin language.
+ */
+export interface Language {
```

Comment:
JS language also has [`visitorKeys`](https://github.com/eslint/eslint/blob/e2b16e2b72606162dce3d804bc80186b6c5ec0f9/lib/languages/js/index.js#L65C5-L65C16) property. Is it intentionally omitted from this interface because ESLint doesn't use it?

Reply:
I was trying to sync between what we were actually doing, but I think ESLint should be updated to use `visitorKeys` on the language object instead of the `SourceCode` object. Then I can add `visitorKeys` here.

Reply:
> ESLint should be updated to use `visitorKeys` on the language object instead of the `SourceCode` object.

That would be problematic for our JS language implementation, since it supports custom parsers and different parsers return different visitor keys.

Reply:
Fair point. I think the problem here is that the JS implementation will be the exception case in this regard. It's unlikely that other languages will allow swapping out of parsers to something that returns an unknown AST.

So probably what we should be doing is `sourceCode.visitorKeys ?? language.visitorKeys` in ESLint, so that there's a way to override `visitorKeys` for a particular file if necessary, but it doesn't require every language to implement `visitorKeys` on `SourceCode`.